src/coreneuron/io/nrn_setup.cpp

```diff
@@ -409,22 +406,22 @@ void nrn_setup(const char* filesdat,
                double* mindelay) {
     double time = nrn_wtime();
 
-    int ngroup;
-    int* gidgroups;
-    nrn_read_filesdat(ngroup, gidgroups, filesdat);
-    UserParams userParams(ngroup,
-                          gidgroups,
-                          datpath,
-                          strlen(restore_path) == 0 ? datpath : restore_path,
-                          checkPoints);
-
+    UserParams userParams(
+        nrn_read_filesdat(filesdat),
+        datpath,
+        strlen(restore_path) == 0 ? datpath : restore_path,
+        checkPoints
+    );
 
     // temporary bug work around. If any process has multiple threads, no
     // process can have a single thread. So, for now, if one thread, make two.
     // Fortunately, empty threads work fine.
     // Allocate NrnThread* nrn_threads of size ngroup (minimum 2)
     // Note that rank with 0 dataset/cellgroup works fine
-    nrn_threads_create(userParams.ngroup <= 1 ? 2 : userParams.ngroup);
+    int n_cell_groups = userParams.cell_groups.size();
+    int n_threads = n_cell_groups <= 1 ? 2 : n_cell_groups;
+    userParams.cell_groups.reserve(n_threads);
```

Comment:
In the case of `n_threads == 1`, what should the value of `???` be:
```
    auto * grp = userParams.cell_groups.data();
    assert grp[1] == ???;
```
IIUC, it used to be `0`. Now I'm not sure it has a value, because we only reserve the memory and IIRC, it's not initialized yet. (It will often be `0`, because the OS zeroes out pages, but I don't think there's a guarantee.)

Reply:
That is a great question! But where do you see that it used to be 0? Happy to set it. But I'm even wondering if we actually had such extended allocation because I realise`iNumFiles / nrnmpi_numprocs + 1` is simply the size required when it's not a perfect division

Reply:
I hate this part of C++, for non-trivial types I know for sure that `new T[N]` will initialize the values allocated. However, `new int[n]` gets the optimized experience.

Reply:
"optimized" 😆 
I got rid of the extra allocation in the last commit. Tests pass, although the nvhpc tests randomly fail. Is that normal?

Reply:
I haven't checked, but NVHPC seems to have an issue with one particular CVODE test, see: https://github.com/neuronsimulator/nrn/issues/3172
If it's exactly that test, I think you can safely ignore it for now.